### DISCUSSION TITLE
How to trigger options validation per tenant on startup?

### DISCUSSION THREAD

**Discussion by xperiandri** [OrchardCMS/OrchardCore#14489] [Q&A]
According to https://github.com/OrchardCMS/OrchardCore/discussions/10229#discussioncomment-1291620 I've moved all the tanant dependant services into
``` C#
.AddOrchardCore()
    .ConfigureServices(svc ->
        ... tenant dependant services
    )
```
then I have options validation that must be triggered on startup
``` F#
    member internal services.``AddPostgreSQL Options``(sectionName, [<Optional>] name) =
        services
            .AddSingleton<IValidateOptions<StorageOptions>, ValidatableOptionValidator<StorageOptions>>()
            .AddOptions<StorageOptions>(name)
            .Configure(fun options (shellConfiguration: IShellConfiguration) -> shellConfiguration.GetSection(StorageOptions.SectionName).Bind options)
            .BindConfiguration(sectionName)
            .ValidateOnStart()
            .Services
```
configuration and validation successfully happens when I access the tenant.
But I want that to happen on startup. And
``` JSON
"OrchardCore_BackgroundService": {
  "ShellWarmup": true
}
```
does not help with that.

What should I modify to achieve my goal of validating setting for all declared tenants on startup?

**Comment by xperiandri** [comment]
@jtkech how to force `ValidationHostedService` to run for each tenant on startup?
https://github.com/dotnet/runtime/blob/42687a9797341f7749ebb7bb316ce3b2f03682e4/src/libraries/Microsoft.Extensions.Hosting/src/OptionsBuilderExtensions.cs
https://github.com/dotnet/runtime/blob/9fc6427cd5f0f9d0f3eb3c2901bfaea80dc2ec18/src/libraries/Microsoft.Extensions.Hosting/src/ValidationHostedService.cs

**Comment by jtkech** [comment]
We can't register an `IHostedService` at the tenant level but there are other ways.

Here 3 ways I have in mind.

- You can use the `ShellWarmup` and define an `IBackgroundTask` to apply your logic.

- Or define a `ModularTenantEvents` and apply your logic in the `ActivatingAsync()` event. But to be run from the beginning you still need the `ShellWarmup` and at least one registered `IBackgroundTask` even if it does nothing.

- Or you can define a service Initializer.

            builder.ConfigureServices(tenantServices =>
            {
                tenantServices.Initialize(sp => // The delegate can be async "async sp =>"
                {
                    var validators = sp.GetRequiredService<IOptions<ValidatorOptions> validatorOptions>();
                    ...
                    ...

                    return Task.CompletedTask;
                });
